features/machine/autoscaler.feature

```diff
@@ -52,3 +52,104 @@ Feature: Cluster Autoscaler Tests
+    Given I use the "openshift-machine-api" project
+    And I pick a random machineset to scale
+
+    When I run oc create over "https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/cloud/machine-autoscaler.yml" replacing paths:
+      | ["metadata"]["name"]                   | maotest                    |
+      | ["spec"]["scaleTargetRef"]["name"]     | <%= machine_set.name %>    |
+    Then the step should succeed
+    And admin ensures "maotest" machineautoscaler is deleted after scenario
+    When I run the :get admin command with:
+      | resource          | machineautoscaler            |
+      | resource_name     | maotest                      |
+    Then the step succeeded
+
+  # @author zhsun@redhat.com
+  # @case_id OCP-22102
+  @admin
+  @destructive
+  Scenario: Update machineAutoscaler to reference a different MachineSet
+    Given I have an IPI deployment
+    And I switch to cluster admin pseudo user
+    Given I use the "openshift-machine-api" project
+    And I store all machinesets in the "openshift-machine-api" project to the :machinesets clipboard
+
+    When I run oc create over "https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/cloud/machine-autoscaler.yml" replacing paths:
+      | ["metadata"]["name"]                   | maotest0                                                    |
+      | ["spec"]["scaleTargetRef"]["name"]     | <%= cb.machinesets[0].name %>                               |
+    Then the step should succeed
+    And admin ensures "maotest0" machineautoscaler is deleted after scenario
+    When I run oc create over "https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/cloud/machine-autoscaler.yml" replacing paths:
+      | ["metadata"]["name"]                   | maotest1                                                    |
+      | ["spec"]["scaleTargetRef"]["name"]     | <%= cb.machinesets[1].name %>                               |
+    Then the step should succeed
+    And admin ensures "maotest1" machineautoscaler is deleted after scenario
+    
+    When I run the :patch admin command with:
+      | resource          | machineautoscaler                                                                |
+      | resource_name     | maotest0                                                                         |
+      | p                 | {"spec":{"scaleTargetRef":{"name":"<%= cb.machinesets[2].name %>"}}}             |
+      | type              | merge                                                                            |
+    Then the step should succeed
+    When I run the :describe admin command with:
+      | resource          | machineautoscaler                                                                |
+      | name              | maotest0                                                                         |
+    Then the step should succeed
+    And the output should match "Name:\s+<%= cb.machinesets[2].name %>"  
```

Comment:
For this test, you assumed there are 3 machinesets in the cluster. Sometimes in an OSP deployment, there is only 1 machineset with 3 machines. The index used here will cause trouble.

Reply:
How about clone 2 machinesets and patch mao for them?

Reply:
I cloned 3 machinesets, if clone 2 machineset, we could delete one mao and test another scenario.

      [16:16:20] INFO> === End After Scenario: CAO listens and annotations machineSets based on MachineAutoscaler resource ===

1 scenario (1 passed)
19 steps (19 passed)
7m2.305s

      [21:10:14] INFO> === End After Scenario: Update machineAutoscaler to reference a different MachineSet ===
1 scenario (1 passed)
42 steps (42 passed)
15m31.919s